tsl/src/compression/arrow_c_data_interface.h

```diff
@@ -0,0 +1,149 @@
+ * necessarily using Arrow libraries or reinventing the wheel.
+ */
+
+#pragma once
+
+#ifndef ARROW_C_DATA_INTERFACE
+#define ARROW_C_DATA_INTERFACE
+
+#define ARROW_FLAG_DICTIONARY_ORDERED 1
+#define ARROW_FLAG_NULLABLE 2
+#define ARROW_FLAG_MAP_KEYS_SORTED 4
+
+typedef struct ArrowArray
+{
+	/*
+	 * Mandatory. The logical length of the array (i.e. its number of items).
+	 */
+	int64 length;
+
+	/*
+	 * Mandatory. The number of null items in the array. MAY be -1 if not yet
+	 * computed.
+	 */
+	int64 null_count;
+
+	/*
+	 * Mandatory. The logical offset inside the array (i.e. the number of
+	 * items from the physical start of the buffers). MUST be 0 or positive.
+	 *
+	 * Producers MAY specify that they will only produce 0-offset arrays to
+	 * ease implementation of consumer code. Consumers MAY decide not to
+	 * support non-0-offset arrays, but they should document this limitation.
+	 */
+	int64 offset;
+
+	/*
+	 * Mandatory. The number of physical buffers backing this array. The
+	 * number of buffers is a function of the data type, as described in the
+	 * Columnar format specification.
+	 *
+	 * Buffers of children arrays are not included.
+	 */
+	int64 n_buffers;
+
+	/*
+	 * Mandatory. The number of children this type has.
+	 */
+	int64 n_children;
+
+	/*
+	 * Mandatory. A C array of pointers to the start of each physical buffer
+	 * backing this array. Each void* pointer is the physical start of a
+	 * contiguous buffer. There must be ArrowArray.n_buffers pointers.
+	 *
+	 * The producer MUST ensure that each contiguous buffer is large enough to
+	 * represent length + offset values encoded according to the Columnar
+	 * format specification.
+	 *
+	 * It is recommended, but not required, that the memory addresses of the
+	 * buffers be aligned at least according to the type of primitive data
+	 * that they contain. Consumers MAY decide not to support unaligned
+	 * memory.
+	 *
+	 * The buffer pointers MAY be null only in two situations:
+	 *
+	 * - for the null bitmap buffer, if ArrowArray.null_count is 0;
+	 *
+	 * - for any buffer, if the size in bytes of the corresponding buffer would
+	 * be 0.
+	 *
+	 * Buffers of children arrays are not included.
+	 */
+	const void **buffers;
+
+	struct ArrowArray **children;
+	struct ArrowArray *dictionary;
+
+	/*
+	 * Mandatory. A pointer to a producer-provided release callback.
+	 *
+	 * See below for memory management and release callback semantics.
+	 */
+	void (*release)(struct ArrowArray *);
+
+	/* Opaque producer-specific data */
+	void *private_data;
+} ArrowArray;
+
+/*
+ * We don't use the schema but have to define it for completeness because we're
+ * defining ARROW_C_DATA_INTERFACE macro.
+ */
+struct ArrowSchema
+{
+	const char *format;
+	const char *name;
+	const char *metadata;
+	int64 flags;
+	int64 n_children;
+	struct ArrowSchema **children;
+	struct ArrowSchema *dictionary;
+
+	void (*release)(struct ArrowSchema *);
+	void *private_data;
+};
+
+#endif
```

Comment:
Is there a specific reason why the include guard ends here and the following functions are not included?

Reply:
Added a comment:
```
/*
 * The include guard ARROW_C_DATA_INTERFACE is required by the Arrow docs to
 * avoid redefinition of the Arrow structs in the third-party headers, but the
 * following functions are not part of Arrow C Data Interface, so they are not
 * under the guard. We still need some kind of guard for them, so we also have
 * pragma once above.
 */
```